akka-stream/src/main/scala/akka/stream/impl/fusing/GraphInterpreter.scala

```diff
@@ -0,0 +1,434 @@
+    else assembly.stages(assembly.inOwners(connection)).toString
+
+  // Debug name for a connections ouput part
+  private def outOwnerName(connection: Int): String =
+    if (assembly.outOwners(connection) == Boundary) "UpstreamBoundary"
+    else assembly.stages(assembly.outOwners(connection)).toString
+
+  /**
+   * Executes pending events until the given limit is met. If there were remaining events, isSuspended will return
+   * true.
+   */
+  def execute(eventLimit: Int): Unit = {
+    var eventsRemaining = eventLimit
+    var connection = dequeue()
+    while (eventsRemaining > 0 && connection != NoEvent) {
+      processEvent(connection)
+      eventsRemaining -= 1
+      if (eventsRemaining > 0) connection = dequeue()
+    }
+    // TODO: deadlock detection
+  }
+
+  // Decodes and processes a single event for the given connection
+  private def processEvent(connection: Int): Unit = {
+
+    def processElement(elem: Any): Unit = {
+      if (!isStageCompleted(assembly.inOwners(connection))) {
+        if (GraphInterpreter.Debug) println(s"PUSH ${outOwnerName(connection)} -> ${inOwnerName(connection)}, $elem")
+        inAvailable(connection) = true
+        inHandlers(connection).onPush()
+      }
+    }
+
+    connectionStates(connection) match {
+      case Pushable ⇒
+        if (!isStageCompleted(assembly.outOwners(connection))) {
+          if (GraphInterpreter.Debug) println(s"PULL ${inOwnerName(connection)} -> ${outOwnerName(connection)}")
+          outAvailable(connection) = true
+          outHandlers(connection).onPull()
+        }
+      case Completed ⇒
+        val stageId = assembly.inOwners(connection)
+        if (!isStageCompleted(stageId)) {
+          if (GraphInterpreter.Debug) println(s"COMPLETE ${outOwnerName(connection)} -> ${inOwnerName(connection)}")
+          inAvailable(connection) = false
+          inHandlers(connection).onUpstreamFinish()
+          completeConnection(stageId)
+        }
+      case Failed(ex) ⇒
+        val stageId = assembly.inOwners(connection)
+        if (!isStageCompleted(stageId)) {
+          if (GraphInterpreter.Debug) println(s"FAIL ${outOwnerName(connection)} -> ${inOwnerName(connection)}")
+          inAvailable(connection) = false
+          inHandlers(connection).onUpstreamFailure(ex)
+          completeConnection(stageId)
+        }
+      case Cancelled ⇒
+        val stageId = assembly.outOwners(connection)
+        if (!isStageCompleted(stageId)) {
+          if (GraphInterpreter.Debug) println(s"CANCEL ${inOwnerName(connection)} -> ${outOwnerName(connection)}")
+          outAvailable(connection) = false
+          outHandlers(connection).onDownstreamFinish()
+          completeConnection(stageId)
+        }
+      case PushCompleted(elem) ⇒
+        inAvailable(connection) = true
+        connectionStates(connection) = elem
+        processElement(elem)
+        enqueue(connection, Completed)
+      case pushedElem ⇒ processElement(pushedElem)
```

Comment:
nope, it will be changed to Empty as soon as grab() is called on that port. Until then, it needs to be kept there.